File: src/profile/ui/gamespecific.cpp
#include "gamespecific.h"
#include "../profile_internal.h"
#include "../../ui/ui.h"

#include "../games/chlcc/delusiontrigger.h"
#include "../games/cclcc/delusiontrigger.h"
#include "../games/cclcc/yesnotrigger.h"
#include "../games/cclcc/mapsystem.h"

namespace Impacto {
namespace Profile {
namespace GameSpecific {

using namespace Impacto::UI;

void Configure() {
  if (!TryPushMember("GameSpecific")) return;
  AssertIs(LUA_TTABLE);
  GameSpecificType =
      GameSpecificType::_from_integral_unchecked(EnsureGetMember<int>("Type"));

  // Get common gamespecific UI elements
  if (GameSpecificType == +GameSpecificType::CCLCC) {
    UIButtonGuideSprites = GetMemberVector<Sprite>("UIButtonGuideSprites");
    UIButtonGuideEndDisp = EnsureGetMember<RectF>("UIButtonGuideEndDisp");
  }
  if (GameSpecificType == +GameSpecificType::CHLCC) {
    MonitorScanline = EnsureGetMember<Sprite>("MonitorScanline");
    EyecatchStar = EnsureGetMember<Sprite>("EyecatchStar");
    ButterflySprites = GetMemberVector<Sprite>("ButterflySprites");
    ButterflyFrameCount = EnsureGetMember<uint8_t>("ButterflyFrameCount");
    ButterflyFlapFrameDuration =
        EnsureGetMember<float>("ButterflyFlapFrameDuration");
    ButterflyFadeDuration = EnsureGetMember<float>("ButterflyFadeDuration");

    BubbleSpriteSmall = EnsureGetMember<Sprite>("BubbleSpriteSmall");
    BubbleSpriteBig = EnsureGetMember<Sprite>("BubbleSpriteBig");
    BubbleFadeDuration = EnsureGetMember<float>("BubbleFadeDuration");
  }

  Pop();

  switch (GameSpecificType) {
    case +GameSpecificType::CHLCC:
      CHLCC::DelusionTrigger::Configure();
      break;
    case +GameSpecificType::CC:
      break;
    case +GameSpecificType::CCLCC:
      CCLCC::MapSystem::Configure();
      CCLCC::YesNoTrigger::Configure();
      CCLCC::DelusionTrigger::Configure();
      break;
    case +GameSpecificType::RNE:
      break;
    case +GameSpecificType::Dash:
      break;
    case +GameSpecificType::None:
      break;
  }
}

}  // namespace GameSpecific
}  // namespace Profile
}  // namespace Impacto

File: profiles/chlcc/gamespecific.lua
root.GameSpecific = {
  Type = GameSpecificType.CHLCC,

  MonitorScanline = "MonitorScanline",

  ButterflySprites = {},
  ButterflyFlapFrameDuration = 4/60,
  ButterflyFrameCount = 8,
  ButterflyFadeDuration = 64 / 60,

  BubbleSpriteSmall = "BubbleSpriteSmall",
  BubbleSpriteBig = "BubbleSpriteBig",
  BubbleFadeDuration = 64/60,

  EyecatchStar = "EyecatchStar",

  MenuSelectPromptDuration =  110 / 60,
  MenuSelectPromptInterval = 5 / 60,
  MenuTransitionDuration = 64 / 60,
  ShowPageAnimationStartTime = 16 / 60,
  ShowPageAnimationDuration = (48 - 16) / 60,
};

root.Sprites["MonitorScanline"] = {
  Sheet = "Data",
  Bounds = { X = 1, Y = 369, Width = 640, Height = 360 },
};

root.Sprites["EyecatchStar"] = {
  Sheet = "Data",
  Bounds = { X = 1, Y = 917, Width = 106, Height = 106 },
};

for i = 0, 5 do
  local name = "ButterflySprites" .. i;
  root.Sprites[name] = {
      Sheet = "Data",
      Bounds = {
          X = i * 100 + 1,
          Y = 817,
          Width = 98,
          Height = 98
      },
  };
  root.GameSpecific.ButterflySprites[i + 1] = name
end

root.Sprites["BubbleSpriteSmall"] = {
    Sheet = "Data",
    Bounds = {
        X = 129,
        Y = 981,
        Width = 22,
        Height = 22
    },
};

root.Sprites["BubbleSpriteBig"] = {
    Sheet = "Data",
    Bounds = {
        X = 179,
        Y = 967,
        Width = 50,
        Height = 50
    },
};

include('chlcc/hud/delusiontrigger.lua');